### PR TITLE
support aes-gcm ciphers in enc app

### PR DESCRIPTION
**Pull request by letslawful**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: apps/enc.c [hunk 1]
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright 1995-2017 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2017 BaishanCloud. All Rights Reserved.
```
Comment (mattcaswell) [inline apps/enc.c]: Please remove this line. We generally don't add an extra copyright line like this unless there is a major contribution (e.g. usually when the file is created)

File: apps/enc.c [hunk 2]
```diff
@@ -268,9 +278,11 @@ int enc_main(int argc, char **argv)
         goto opthelp;
     }
 
-    if (cipher && EVP_CIPHER_flags(cipher) & EVP_CIPH_FLAG_AEAD_CIPHER) {
-        BIO_printf(bio_err, "%s: AEAD ciphers not supported\n", prog);
-        goto end;
+    if (cipher && (EVP_CIPHER_flags(cipher) & EVP_CIPH_FLAG_AEAD_CIPHER)) {
+        if ((EVP_CIPHER_mode(cipher) != EVP_CIPH_GCM_MODE)) {
```
Comment (mattcaswell) [inline apps/enc.c]: Why are you restricting this to GCM only? It seems that it would not be too difficult to make this generic.

File: apps/enc.c [hunk 3]
```diff
@@ -347,6 +360,31 @@ int enc_main(int argc, char **argv)
...
+#ifndef OPENSSL_NO_UI
+        for (;;) {
+            char prompt[200];
+
+            BIO_snprintf(prompt, sizeof prompt, "enter %s %s aad:",
+                         OBJ_nid2ln(EVP_CIPHER_nid(cipher)),
+                         (enc) ? "encryption" : "decryption");
+            aadbuf[0] = '\0';
+            i = EVP_read_aad_string((char *)aadbuf, SIZE, prompt);
+            if (i == 0) {
```
Comment (mattcaswell) [inline apps/enc.c]: So EVP_read_aad_string() returns 0 on success? That is contrary to our normal way of doing things (1 is success, 0 (and sometimes -1) is error). Mind you the UI APIs are slightly different so I defer to @levitte on the right way to do this.

File: doc/man1/enc.pod [hunk 4]
```diff
@@ -98,6 +99,10 @@ If the B<-a> option is set then base64 process the data on one line.
 The password to derive the key from. This is for compatibility with previous
 versions of OpenSSL. Superseded by the B<-pass> argument.
 
+=item B<-aad arg>
+
+specify the additional authenticated data for AEAD ciphers.
```
Comment (InfoHunter) [inline doc/man1/enc.pod]: s/specify/Specify

File: apps/enc.c [hunk 5]
```diff
@@ -512,18 +574,45 @@ int enc_main(int argc, char **argv)
...
+                printf("aad=%s", aad);
+                printf("\n");
+            }
             if (!nosalt) {
                 printf("salt=");
                 for (i = 0; i < (int)sizeof(salt); i++)
                     printf("%02X", salt[i]);
                 printf("\n");
             }
+            if (tag[0]) {
```
Comment (mattcaswell) [inline apps/enc.c]: What happens if the first byte of a tag is 0?

File: apps/enc.c (hunk 1 above)
Comment (InfoHunter) [inline apps/enc.c]: This should be removed...

File: include/openssl/evp.h [hunk 6]
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright 1995-2017 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2017 BaishanCloud. All Rights Reserved.
```
Comment (InfoHunter) [inline include/openssl/evp.h]: This should be removed...

File: doc/man1/enc.pod [hunk 7]
```diff
@@ -316,6 +319,7 @@ authentication tag.
  aes-[128|192|256]-ctr  128/192/256 bit AES in CTR mode
  aes-[128|192|256]-ecb  128/192/256 bit AES in ECB mode
  aes-[128|192|256]-ofb  128/192/256 bit AES in OFB mode
+ aes-[128|192|256]-ofb  128/192/256 bit AES in GCM mode
```
Comment (InfoHunter) [inline doc/man1/enc.pod]: aes-[128|192|256]-gcm ?

File: apps/enc.c [hunk 8]
```diff
@@ -512,18 +574,45 @@ int enc_main(int argc, char **argv)
             goto end;
         }
 
+        if ((EVP_CIPHER_flags(cipher) & EVP_CIPH_FLAG_AEAD_CIPHER)) {
+            if (!EVP_CipherUpdate(ctx, NULL, &outlen, (unsigned char*)aad, strlen(aad))) {
```
Comment (mattcaswell) [inline apps/enc.c]: Shouldn't we include the magic/salt etc into the AAD? i.e. everything in the file except the tag itself should be covered somehow by the integrity guarantee.

File: include/openssl/evp.h (hunk 6 above)
Comment (mattcaswell) [inline include/openssl/evp.h]: As above

File: doc/man1/enc.pod (hunk 7 above)
Comment (letslawful) [reply doc/man1/enc.pod]: yes, a typo here.

File: apps/enc.c [hunk 9]
```diff
@@ -68,6 +68,7 @@ const OPTIONS enc_options[] = {
      "Used with -[base64|a] to specify base64 buffer as a single line"},
     {"bufsize", OPT_BUFSIZE, 's', "Buffer size"},
     {"k", OPT_K, 's', "Passphrase"},
+    {"aad", OPT_AAD, 's', "Additional Authenticated Data for AEAD ciphers"},
```
Comment (paulidale) [inline apps/enc.c]: Is `addfile` worth considering as well?

File: apps/enc.c (hunk 2 above)
Comment (paulidale) [reply apps/enc.c]: I'm with @mattcaswell here, why not CCM?

**Comment by dot-asm** [comment]
CTR is not really suitable mode for storage, GCM is really a communication-oriented mode. It makes much more sense to invest into support for mode more suitable for storage, even specifically designed for it, XTS.

**Comment by richsalz** [comment]
I think the issues that @dot-asm and @paulidale raised are important.  Perhaps we should discuss this on the openssl-dev mailing list?

File: apps/enc.c (hunk 2 above)
Comment (zakjan) [reply apps/enc.c]: CCM tests fail in my alternative implementation #5026, maybe that's the reason? Other ciphers work fine.

Comment (zakjan) [reply apps/enc.c]: CCM doesn't support streaming https://github.com/openssl/openssl/pull/5026#issuecomment-355779385

**Comment by richsalz** [comment]
There is also discussion going on in #5026.

**Comment by dot-asm** [comment]
As per #5048 this is closing without resolution.

**Pull request closed without merging**